### PR TITLE
Add test for vBalance function

### PR DESCRIPTION
**Pull request by eric6311**
- fix  cannot  get `totalvBalanceAtTime()` in future
- add test for each vBalance and totalVBalance
- add event assertion test for Withdraw()

### CODE REVIEW

**Review by zeqing-guo — APPROVED** [review]

File: contracts/test/veLON/vBalance.t.sol [hunk 1]
```diff
@@ -53,4 +53,97 @@ contract TestVeLONDeposit is TestVeLON {
...
+    function testVBalanceOfAtTime() public {
+        uint256 aliceStakeAmount = 10 * 365 days;
+        uint256 aliceTokenId = _stakeAndValidate(alice, aliceStakeAmount, 2 weeks);
+
+        // fast forward 1 week
+        vm.warp(block.timestamp + 1 weeks);
+        vm.roll(block.number + 1);
+
+        // calculate Alice's balance
+        uint256 expectAliceBalance = 10 * 7 days;
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:76]: Add comments of how the calculation is done.

File: contracts/test/veLON/vBalance.t.sol [hunk 2]
```diff
@@ -53,4 +53,97 @@ contract TestVeLONDeposit is TestVeLON {
...
+        vm.expectRevert("Invalid block number");
+        // get balance in future block
+        veLon.vBalanceOfAtBlk(tokenId, block.number + 1);
+    }
+
+    /***********************************
+     *   Uint Test for Total Balance *
+     ***********************************/
+    function testTotalvBalance() public {
+        _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, 2 weeks);
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:107]: Add another user and stake, check if the total is equal to sum(s1, s2).

File: contracts/test/veLON/vBalance.t.sol [hunk 3]
```diff
@@ -53,4 +53,97 @@ contract TestVeLONDeposit is TestVeLON {
...
+    }
+
+    function testTotalvBalanceAtBlk() public {
+        _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, 2 weeks);
+        uint256 totalvBalance = veLon.totalvBalanceAtBlk(block.number);
+        uint256 expectvBalance = 10 * 2 weeks;
+        assertEq(totalvBalance, expectvBalance);
+
+        // next block and 1 week has passed
+        expectvBalance = 10 * 1 weeks;
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:139]: Perhaps 10 blocks and 1 week? Or other combination to test the block ratio calculation.

File: contracts/test/veLON/Withdraw.t.sol [hunk 4]
```diff
@@ -35,10 +42,14 @@ contract TestVeLONWithfraw is TestVeLON {
...
+
+        // calculate the panalty
+        uint256 expectedPanalty = ((DEFAULT_STAKE_AMOUNT.mul(veLon.earlyWithdrawPenaltyRate())).div(veLon.PENALTY_RATE_PRECISION()));
+        uint256 expectedAmount = DEFAULT_STAKE_AMOUNT.sub(expectedPanalty);
+        vm.expectEmit(true, true, true, true);
+        emit Withdraw(user, false, tokenId, expectedAmount, expectedPanalty, block.timestamp);
         vm.prank(user);
         veLon.withdrawEarly(tokenId);
-        uint256 balanceChange = DEFAULT_STAKE_AMOUNT.mul(earlyWithdrawPenaltyRate).div(PENALTY_RATE_PRECISION);
-        stakerLon.assertChange(int256(DEFAULT_STAKE_AMOUNT.sub(balanceChange)));
```
Comment (charlesjhongc) [inline contracts/test/veLON/Withdraw.t.sol:41]: This assertion should not be remove. Also, `lockedLon` should be checked as well.

Comment (eric6311) [reply contracts/test/veLON/Withdraw.t.sol:41]: I originally thinked that the balance change has tested in `expectEmit withdraw` , so  I removed these two line. But I can add them back.

Comment (charlesjhongc) [reply contracts/test/veLON/Withdraw.t.sol:41]: Considering them as the same is an assumption. Let's verify it as much as you can.

File: contracts/test/veLON/vBalance.t.sol [hunk 5]
```diff
@@ -54,6 +55,42 @@ contract TestVeLONDeposit is TestVeLON {
...
+        emit log(Strings.toString(veLon.vBalanceOf(bobTokenId)));
+        emit log(Strings.toString(veLon.vBalanceOf(aliceTokenId)));
+        emit log(Strings.toString(veLon.vBalanceOfAtTime(aliceTokenId, block.timestamp)));
+        // sub Bob's all balance
+        totalBalance = totalBalance.sub(initialBobBalance);
+        // sub Alice's delined balance ()
+        totalBalance = totalBalance.sub((1 * 7 days));
+        vm.prank(bob);
+        veLon.withdraw(bobTokenId);
+        assertEq(veLon.totalvBalance(), totalBalance, "expectedBalance is not equal");
```
Comment (eric6311) [inline contracts/test/veLON/vBalance.t.sol:91]: the expected totalBalance is **604800**, but actual is **604799**

Comment (eric6311) [inline contracts/test/veLON/vBalance.t.sol:91]: In the last assertion , the expected totalBalance is **604800**, but actual is **604799**

File: contracts/test/veLON/vBalance.t.sol [hunk 6]
```diff
@@ -54,6 +55,42 @@ contract TestVeLONDeposit is TestVeLON {
...
+        totalBalance = totalBalance.add(initialBobBalance);
+        assertEq(veLon.vBalanceOf(bobTokenId), initialBobBalance);
+        assertEq(veLon.totalvBalance(), totalBalance);
+
+        // check the totalSupply
+        assertEq(veLon.totalSupply(), 2);
+
+        // fast forward 1 weeks, bob's lock is expired
+        uint256 dt = 1 days;
+        vm.roll(block.timestamp + 7 * dt);
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:80]: `vm.roll` is used to set block number while `vm.warp` is for block timestamp. I got case passed after swapping these two.

File: contracts/test/veLON/vBalance.t.sol [hunk 7]
```diff
@@ -53,4 +54,222 @@ contract TestVeLONDeposit is TestVeLON {
...
+        uint256 expectedAliceBalance = 1 weeks;
+        assertEq(veLon.vBalanceOf(aliceTokenId), expectedAliceBalance, "Alice's balance is not as expected");
+
+        vm.prank(bob);
+        veLon.withdraw(bobTokenId);
+        assertEq(veLon.totalvBalance(), totalBalance, "expectedBalance is not equal");
+    }
+
+    function testCreateLock_and_OneEarlyWithdraw() public {
+        // alice and bob despoit for different amount and time
```
Comment (charlesjhongc) [inline contracts/test/veLON/vBalance.t.sol:97]: Both use the same amount and lock duration. Update comment or change testing code.

File: contracts/test/veLON/vBalance.t.sol [hunk 8]
```diff
@@ -2,9 +2,10 @@
 pragma solidity 0.7.6;
 
 import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
```
[the remark this answers is not in the document]
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol]: Sorry, removed in [96afd71](https://github.com/consenlabs/tokenlon-contracts/pull/72/commits/96afd714ea7546409f3b4d9755be9e6a999db63e)

File: contracts/test/veLON/vBalance.t.sol [hunk 9]
```diff
@@ -53,4 +54,222 @@ contract TestVeLONDeposit is TestVeLON {
...
+        vm.warp(block.timestamp + 7 * dt);
+        vm.roll(block.number + 1);
+        // sub Bob's all balance
+        totalBalance = totalBalance.sub(bobBalance);
+        // sub Alice's delined balance ()
+        totalBalance = totalBalance.sub((1 * 7 days));
+
+        // alice's balance
+        uint256 expectedAliceBalance = 1 weeks;
+        assertEq(veLon.vBalanceOf(aliceTokenId), expectedAliceBalance, "Alice's balance is not as expected");
```
[the remark this answers is not in the document]
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol:88]: added in [16b33ab](https://github.com/consenlabs/tokenlon-contracts/pull/72/commits/16b33ab47ff9bdbdf0c1ccf7b133867fdcd92f26)

File: contracts/test/veLON/vBalance.t.sol [hunk 10]
```diff
@@ -53,4 +54,222 @@ contract TestVeLONDeposit is TestVeLON {
...
+
+        vm.prank(bob);
+        veLon.withdraw(bobTokenId);
```
[the remark this answers is not in the document]
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol:98]: added in [16b33ab](https://github.com/consenlabs/tokenlon-contracts/pull/72/commits/16b33ab47ff9bdbdf0c1ccf7b133867fdcd92f26)

File: contracts/test/veLON/vBalance.t.sol (hunk 7 above)
Comment (eric6311) [reply contracts/test/veLON/vBalance.t.sol]: Updated the comments for **same amount and time**

**Review by charlesjhongc — APPROVED** [review]

**Pull request merged**